File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/app/receipt/revision/item/ItemEntity.java
package pl.madzierski.daniel.app.receipt.revision.item;

import jakarta.persistence.*;
import lombok.*;
import pl.madzierski.daniel.app.common.model.BaseEntity;
import pl.madzierski.daniel.app.receipt.revision.ReceiptRevisionEntity;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "receipt_item")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ItemEntity extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    private ReceiptRevisionEntity receiptRevision;

    private String name;

    private Double amount;

    private Double unitPrice;

    private Double discount;

    private Double totalPrice;

    private Integer position;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_item_id")
    private ItemEntity parentItem;

    @OneToMany(mappedBy = "parentItem", fetch = FetchType.LAZY)
    private Set<ItemEntity> childItems = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        ItemEntity that = (ItemEntity) o;
        return Objects.equals(name, that.name) && Objects.equals(amount, that.amount) && Objects.equals(unitPrice, that.unitPrice) && Objects.equals(discount, that.discount) && Objects.equals(totalPrice, that.totalPrice) && Objects.equals(position, that.position);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, amount, unitPrice, discount, totalPrice, position);
    }
}

File: receipt-analyzer-service/src/main/kotlin/pl/madzierski/daniel/app/receipt/revision/item/ItemEntity.kt
package pl.madzierski.daniel.app.receipt.revision.item

import jakarta.persistence.*
import pl.madzierski.daniel.app.common.model.BaseEntity
import pl.madzierski.daniel.app.receipt.revision.ReceiptRevisionEntity

@Entity
@Table(name = "receipt_item")
data class ItemEntity(

    @ManyToOne(fetch = FetchType.LAZY)
    var receiptRevision: ReceiptRevisionEntity?,

    var name: String?,

    var ptu: String?,

    var amount: Double?,

    var unitPrice: Double?,

    var discount: Double?,

    var totalPrice: Double?,

    var position: Int?,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_item_id")
    var parentItem: ItemEntity? = null,

    @OneToMany(mappedBy = "parentItem")
    val childItems: MutableSet<ItemEntity> = mutableSetOf()


) : BaseEntity() {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ItemEntity

        if (amount != other.amount) return false
        if (unitPrice != other.unitPrice) return false
        if (discount != other.discount) return false
        if (totalPrice != other.totalPrice) return false
        if (name != other.name) return false
        if (ptu != other.ptu) return false
        if (position != other.position) return false

        return true
    }

    override fun hashCode(): Int {
        var result = amount?.hashCode() ?: 0
        result = 31 * result + (unitPrice?.hashCode() ?: 0)
        result = 31 * result + (discount?.hashCode() ?: 0)
        result = 31 * result + (totalPrice?.hashCode() ?: 0)
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + (ptu?.hashCode() ?: 0)
        result = 31 * result + (position?.hashCode() ?: 0)
        return result
    }

}

